src/vscode-dts/vscode.proposed.debugDataBreakpoints.d.ts

```diff
@@ -0,0 +1,96 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+declare module 'vscode' {
+
+	// https://github.com/microsoft/vscode/issues/195151
+
+	/**
+	 * The source for a data breakpoint.
+	 */
+	export type DataBreakpointSource =
+		| {
+			/** The source type for fixed data identifiers that do not need to be re-resolved when the breakpoint is activated. */
+			type: 'variable';
+			/** An identifier for the data. If it was retrieved using a `variablesReference` it may only be valid in the current suspended state, otherwise it's valid indefinitely. */
+			dataId: string;
+		}
+		| {
+			/** The source type for address-based data breakpoints. Address-based data breakpoints are re-resolved when the breakpoint is activated. This type only applies to sessions that have the `supportsDataBreakpointBytes` capability. */
+			type: 'address';
+			/** A memory address as a decimal value, or hex value if it is prefixed with `0x`. */
+			address: string;
+			/** If specified, returns information for the range of memory extending `bytes` number of bytes from the address. */
+			bytes?: number;
+		}
+		| {
+			/** The source type for expressions that are dynamically resolved when the breakpoint is activated. */
+			type: 'expression';
+			/** A global expression that is first evaluated when the breakpoint is activated. */
+			expression: string;
+		}
+		| {
+			/** The source type for scoped variables and expressions that are dynamically resolved when the breakpoint is activated. */
+			type: 'scoped';
+		} & (
+			| {
+				/** The name of the variable that is used for resolution. */
+				variable: string;
+				/** Reference to the variable container that has the variable named `variable`. */
+				variablesReference: number;
+				frameId?: never;
+			}
+			| {
+				/** The name of the expression that is used for resolution. */
+				expression: string;
+				/** Reference to the stack frame to which the expression is scoped. */
+				frameId: number;
+				variablesReference?: never;
+			}
+		);
```

Comment:
These types are more complex than we generally like in our API. We'd probably represent these as classes that extensions could construct and pass to `DataBreakpointAccessType`

Reply:
I understand, I refactored the types as classes that can be constructed more naturally.